Testnet/MintableTokenInvoice/MintableTokenInvoice/MintableTokenInvoice.cs

```diff
@@ -91,61 +128,114 @@ public string GetInvoiceReference(string transactionReference)
+        // If the invoice is not authorized then return a message to the user.
+        // We don't assert because we want to make the invoice available for approval and allow the user to resubmit the invoice.
+        // We don't want to return the transaction reference until the invoice is authorized.
+        if (!invoice.IsAuthorized)
+        {
+            return $"Obtain authorization for this invoice ({invoiceReference}) then resubmit this request.";
+        }
 
         // Only provide the transaction reference if all checks pass.
         return transactionReference;
     }
 
+    /// <inheritdoc />
+    public string CreateInvoice(string symbol, UInt256 amount, UInt256 fee, UInt128 uniqueNumber, string targetAddress, string targetNetwork)
+    {
+        return CreateInvoiceInternal(Message.Sender, symbol, amount, fee, uniqueNumber, targetAddress, targetNetwork);
+    }
+
+    public string CreateInvoiceFor(Address address, string symbol, UInt256 amount, UInt256 fee, UInt128 uniqueNumber, string targetAddress, string targetNetwork, byte[] signature)
+    {
+        Assert(fee == CalculateFee(amount), "The provided to fee does not match the expected fee.");
+
+        // E.g. for amount 110 and fee = 10 we mint 110 tokens and keep 10 minted tokens as fee.
+        var template = new SignatureTemplate() { UniqueNumber = uniqueNumber, Amount = amount, Fee = fee, Symbol = symbol, TargetAddress = targetAddress, TargetNetwork = targetNetwork, Contract = this.Address };
+        var res = Serializer.Serialize(template);
+        Assert(ECRecover.TryGetSigner(res, signature, out Address signer), "Could not resolve signer.");
+        Assert(signer == address, "Invalid signature.");
+
+        return CreateInvoiceInternal(address, symbol, amount, fee, uniqueNumber, targetAddress, targetNetwork);
+    }
+    
+    public string CreateInvoiceFromURL(Address address, string url, byte[] signature)
+    {
+        Assert(SSAS.TryGetSignerSHA256(Serializer.Serialize(url), signature, out Address signer), "Could not resolve signer.");
+        Assert(signer == address, "Invalid signature.");
+
+        var args = SSAS.GetURLArguments(url, new string[] { "uid", "symbol", "amount", "fee", "targetAddress", "targetNetwork", "contract" });
+
+        Assert(args != null && args.Length == 7, "Invalid url.");
+        Assert(Serializer.ToAddress(SSAS.ParseAddress(args[6], out _)) == this.Address, "Invalid contract address.");
+
+        // Parse amount.
+        string amount = args[2];
+        int amountDecimalIndex = amount.IndexOf('.');
+        int amountDecimals = amountDecimalIndex >= 0 ? amount.Length - amountDecimalIndex - 1 : 0;
+        Assert(amountDecimals <= 2, "Too many decimals in amount");
+
+        var amountUInt256 = UInt256.Parse(amount.PadRight(amount.Length + 8 - amountDecimals, '0').Replace(".", ""));
+
+        // Parse fee.
+        string fee = args[2];
```

Comment:
amount parsing and fee parsing logic is duplicated and it can be moved to a method.

Reply:
Added method.